### PR TITLE
Init Deserialize provider

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

Needs #781 first

Not used and no DI for now, cause we need the ReadProvider first.

### CODE REVIEW

File: src/Component/src/State/Provider/DeserializeProvider.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        private ?SerializerInterface $serializer,
+    ) {
+    }
+
+    public function provide(Operation $operation, Context $context): object|array|null
+    {
+        $data = $this->decorated->provide($operation, $context);
+
+        // We need request content
+        if (!$operation instanceof HttpOperation || !($request = $context->get(RequestOption::class)?->request() ?? null)) {
```
Comment (diimpp) [inline src/Component/src/State/Provider/DeserializeProvider.php]: `$request` is defined conditionally, but used later in the code. I assume conditions defined in such way, that there will be no errors and phpstan didn't raise an issue, but it's a bit hard to read.

As idea, though I don't insist. 
```php
if (!$operation instanceof HttpOperation) {
    return $data;
}

$request = $context->get(RequestOption::class)?->request() ?? null;
if (!request) {
    return $data;
}
```

File: src/Component/src/State/Provider/DeserializeProvider.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        private ProviderInterface $decorated,
+        private ?SerializerInterface $serializer,
+    ) {
+    }
+
+    public function provide(Operation $operation, Context $context): object|array|null
+    {
+        $data = $this->decorated->provide($operation, $context);
+
+        // We need request content
```
Comment (diimpp) [inline src/Component/src/State/Provider/DeserializeProvider.php]: This comment message doesn't help me understand anything, so I would vote for removal. 

Overall it's easy to read that provider checks if operation is deserializeable and either returns inner provider data or deserialize it.

File: src/Component/Tests/State/Provider/DeserializeProviderTest.php [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+use Symfony\Component\HttpFoundation\ParameterBag;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Serializer\SerializerInterface;
+
+final class DeserializeProviderTest extends TestCase
+{
+    use ProphecyTrait;
+
+    /** @var ProviderInterface|ObjectProphecy */
+    private $decorated;
```
Comment (diimpp) [inline src/Component/Tests/State/Provider/DeserializeProviderTest.php]:
```suggestion
    private ProviderInterface|ObjectProphecy $decorated;
```

File: src/Component/Tests/State/Provider/DeserializeProviderTest.php [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+    /** @var SerializerInterface|ObjectProphecy */
+    private $serializer;
```
Comment (diimpp) [inline src/Component/Tests/State/Provider/DeserializeProviderTest.php]:
```suggestion
    private SerializerInterface|ObjectProphecy $serializer;
```

File: src/Component/Tests/State/Provider/DeserializeProviderTest.php [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+    /** @var DeserializeProvider */
+    private $deserializableProvider;
```
Comment (diimpp) [inline src/Component/Tests/State/Provider/DeserializeProviderTest.php]:
```suggestion
    private DeserializeProvider $deserializableProvider;
```

File: src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\Serializer\SerializerInterface;
+
+final class DeserializeListenerTest extends TestCase
+{
+    use ProphecyTrait;
+
+    /** @var HttpOperationInitiatorInterface|ObjectProphecy */
+    private $operationInitiator;
```
Comment (diimpp) [inline src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php]:
```suggestion
    private HttpOperationInitiatorInterface|ObjectProphecy $operationInitiator;
```

File: src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+    /** @var SerializerInterface|ObjectProphecy */
+    private $serializer;
```
Comment (diimpp) [inline src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php]:
```suggestion
    private SerializerInterface|ObjectProphecy $serializer;
```

File: src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    /** @var DeserializeListener */
+    private $deserializableListener;
```
Comment (diimpp) [inline src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php]:
```suggestion
    private DeserializeListener $deserializableListener;
```

File: src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php (hunk 6 above)
Comment (loic425) [reply src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php]: yeah it was on previous PR, and this is changed

File: src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php (hunk 7 above)
Comment (loic425) [reply src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php]: yeah it was on previous PR, and this is changed

File: src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php (hunk 8 above)
Comment (loic425) [reply src/Component/Tests/Symfony/EventListener/DeserializeListenerTest.php]: see comments above

File: src/Component/src/State/Provider/DeserializeProvider.php (hunk 1 above)
Comment (loic425) [reply src/Component/src/State/Provider/DeserializeProvider.php]: done

File: src/Component/Tests/State/Provider/DeserializeProviderTest.php (hunk 3 above)
Comment (diimpp) [reply src/Component/Tests/State/Provider/DeserializeProviderTest.php]: @loic425 I think something happened at rebase, property type is not used here and on other properties

**Review by diimpp — APPROVED** [review]

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thank you, @loic425!

**Pull request merged**